src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java

```diff
@@ -200,6 +205,31 @@ public void complete()
             for (PerIndexWriter perIndexWriter : perIndexWriters)
             {
                 perIndexWriter.complete(stopwatch);
+
+                // The handling of components when we flush is a tad backward: instead of registering the components as
+                // we write them, all the components are collected beforehand in `SSTableWriter#create`, which means
+                // this is a superset of possible components, but if any components are not written for those reason,
+                // this needs to be fixed afterward. One case for SAI component for instance is empty indexes: if a
+                // particular sstable has nothing indexed for a particular index, then only the completion marker for
+                // that index is kept on disk but no other components, so we need to remove the components that were
+                // "optimistically" added (and more generally, future index implementation may have some components that
+                // are only optionally present based on specific flush conditions).
+                // Note: as hinted above, an alternative here would be to change the whole handling of components on
+                // flush, registering components only as they are effectively written. This is a larger refactor, with
+                // some subtleties involved, so it is left as potential future work.
+                if (opType == OperationType.FLUSH)
+                {
+                    var writtenComponents = perIndexWriter.writtenComponents().allAsCustomComponents();
+                    var registeredComponents = IndexDescriptor.perIndexComponentsForNewlyFlushedSSTable(perIndexWriter.indexContext());
+                    var toRemove = Sets.difference(registeredComponents, writtenComponents);
+                    if (!toRemove.isEmpty())
```

Comment:
what about adding some "trace" logging (disabled by default but we can enable it without restarting) ?

Reply:
Fine by me.